### DISCUSSION TITLE
Dealing with overlapping overloads and protocol hierarchies

### DISCUSSION THREAD

**Discussion by rsokl** [microsoft/pyright#3185] [Q&A] [answered]
Hello!

I am overloading a function using types that live in a hierarchy of protocols. pyright ultimately reveals the correct type, but I cannot figure out how to write my overloads without getting reports of overlapping overloads.

Here is a representative example.

```python
from typing_extensions import Protocol, runtime_checkable

@runtime_checkable
class HasA(Protocol):
    a: int

@runtime_checkable
class HasAB(HasA, Protocol):
    b: str


@overload
def get_b_but_a_is_ok_too(x: HasAB) -> str:  # Overload 1 for "get_b_but_a_is_ok_too" overlaps overload 2 and returns an incompatible type
    ...

@overload
def get_b_but_a_is_ok_too(x: HasA) -> int:
    ...


def get_b_but_a_is_ok_too(x: Union[HasAB, HasA]) -> Union[str, int]:
    if isinstance(x, HasAB):
        return x.b
    return x.a


class A:
    a: int = 1

class AB(A):
    b: str = "hi"


a = A()
ab = AB()

reveal_type(get_b_but_a_is_ok_too(a), expected_text="int")  # OK
reveal_type(get_b_but_a_is_ok_too(ab), expected_text="str")  # OK
```

Any guidance here would be much appreciated!

**Comment by erictraut** [comment, marked as answer]
I don't see a good way to avoid an overlapping overload here. The problem is that you've defined two protocols that largely overlap, so any class that satisfies `HasA` will also satisfy `HasAB`. This can lead to incorrect overload selection, so pyright is correct in warning you about the overlap.

You would need to provide some way to discriminate between classes that satisfy `HasA` versus `HasAB`. For example, here I've added an `id` parameter that distinguishes between them in an unambiguous way.

```python
@runtime_checkable
class HasA(Protocol):
    id: Literal["__a__"]
    a: int


@runtime_checkable
class HasAB(Protocol):
    id: Literal["__ab__"]
    a: int
    b: str
```

**Reply by rsokl** [reply]
That makes sense. Thank you very much!
